Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -825,6 +831,24 @@ public struct ByteBuffer {
         }
         return true
     }
+    
+    /// Reset capacity of the buffer to the specified amount.
+    ///
+    /// - returns: `true` if one or more bytes have been discarded, `false` if there are no bytes to discard.
+    @inlinable
+    @discardableResult public mutating func clampBufferCapacity(to maxRetainedCapacity: Int) -> Bool {
```

Comment:
This method is pretty scary to me, as it behaves very strangely if there are more readable bytes in the buffer than the requested capacity. This behaviour is also out-of-line with the comment: the comment says it returns true if bytes have been discarded, but it returns true even if those bytes are not in the readable slice.

I think it might be better to have this refuse to shrink if the target capacity is _smaller_ than the number of readable bytes.

Reply:
"Refuse to shrink if the target capacity is smaller than the number of readable byes"

Wouldn't the target capacity ideally be smaller than the number of readable bytes in the situation that we want to shrink the size of the buffer? 

I think I need some extra help understanding this comment or what the resulting guard statement should look like.

Reply:
The problem is if you would shrink below the number of readable bytes you would basically discard data which is lost. This isn't great and can lead to pretty exciting bugs.
Going back to the original issue. The problem is that the `MessageToByteHandler` might have to write a single large message into the buffer. This buffer is then sent to the pipeline to be written out. This will result in the written bytes of the buffer to be "read" by the pipeline and the `readableBytes` will become `0`. The problem we are trying to solve here is that the buffer is then re-used by the `MessageToByteHandler` for only small messages going forward but it will forever reserve the capacity needed for the single large message.

The flow is basically like this
- Encode large message in `MessageToByteHandler` to `ByteBuffer` `b1`
- Write `b1` in channel pipeline -> `b1.readableBytes` becomes `0`
- Encode small message in `MessageToByteHandler` to `b1` (`b1` is still sized for the previous large message)

What we need here is a clamp that does `max(self.readableBytes, desiredCapacity)`

Reply:
Thank you. This helps me quite a bit. Will continue and push a PR once my second question below is answered about the API initialiser design.